### ISSUE TITLE
RFC: Add support for user defined part/partition data skipping indices

### ISSUE THREAD

**Issue by josh-hildred** [ClickHouse/ClickHouse#65413]
### Proposal 

Allow users to define general lightweight data skipping indices at the part/partition level, rather than just at the granule level.

### Background

ClickHouse currently supports data skipping indexes defined over arbitrary columns at the granule level. These indexes allow ClickHouse to filter granules based on arbitrary columns (rather than just the primary key columns). However, these indexes are evaluated last and can be costly in some scenarios.

When a table is partitioned, ClickHouse automatically adds a MinMax index on the partition key to facilitate pruning partitions before the primary index is used. However, this index cannot be changed (i.e. be defined over a different column). 

### Details

The proposed change would provide users the option to define a data skipping index, at the partition level, over any column and index type. The proposed index would be like the current granule level data skipping index, but at the part/partition level like the MinMax index that is automatically defined for partitions. These indexes would offer more flexibility to filter parts/partitions than existing indices, as they would be definable over any column and any index type (i.e. bloom filter, MinMax, etc.). 

These part/partition level data skipping indexes: a) would be evaluated before the primary index is evaluated, allowing more parts (and thus more granules) to be filtered earlier (leading to less work when the primary index is applied), and b) would be lightweight, as the index would only contain an entry for each part/partition, meaning the index would be of size O(# of parts/partitions).

### Use cases
 
The proposed index would be useful when columns are correlated with the partition key columns. For example, by defining such an index, if a query contains a predicate on the indexed column, but not on the partition key column, ClickHouse would be able to use the index to filter partitions with irrelevant data based on the correlation.  

Secondly, the proposed index could be used in conjunction with granule level data skipping indexes and/or the primary index for better performance. Namely, if both indexes are defined on the same column the part/partition level data skipping index could be used to prefilter parts/partitions, leading to less work when evaluating the corresponding granule level data skipping index and/or primary index.

**Comment by UnamedRus** [comment]
Duplicate 

https://github.com/ClickHouse/ClickHouse/issues/33583

**Comment by josh-hildred** [comment]
@UnamedRus I think both solutions would solve similar problems, but there are differences in what I propose. https://github.com/ClickHouse/ClickHouse/issues/33583 proposes allowing the user to change the order that ClickHouse uses indices (including the primary index). Whereas I proposes adding a completely new index that is always defined at the part/partition level and is used at the same time as the MinMax index over the partitioning key (In `MergeTreeDataSelectExecutor::filterPartsByPartition`). This index would only be available when partitioning.

**Comment by CurtizJ** [comment]
Also see #64210. I think the idea is pretty similar.

**Comment by josh-hildred** [comment]
@CurtizJ Yes, I agree the idea is also similar. I think all three issues agree on the fact that ClickHouse performance could be improved by allowing user defined indices to be used earlier (I.e. to filter parts or partitions) when reading data. Some discussion on what a) makes the most sense for ClickHouse and b) would be most useful to users would be interesting.
